TASK: Prevent multiple catchup runs

kitsunet opened the pull request:
Will use caches to try avoiding to even start async catchups if for that projection one is still running. 
Also adds a simple single slot queue with incremental backup to ensure a requested catchup definitely happens.

kitsunet committed:
TASK: Prevent multiple catchup runs

Will use caches to try avoiding to even start async
catchups if for that projection one is still running.
Also adds a simple single slot queue with incremental
backup to ensure a requested catchup definitely happens.

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Factory/ProjectionCatchUpTrigger/SubprocessProjectionCatchUpTrigger.php:
```diff
@@ -29,15 +38,80 @@ public function triggerCatchUp(Projections $projections): void
-                    'projectionClassName' => get_class($projection)
-                ]
-            );
+        $queuedProjections = array_map(fn($projection) => $this->startCatchUpWithQueueing($projection), iterator_to_array($projections->getIterator()));
```
This could be changed to just
```suggestion
        $queuedProjections = array_map($this->startCatchUpWithQueueing(...), iterator_to_array($projections));
```

kitsunet replied:
good point!

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Factory/ProjectionCatchUpTrigger/SubprocessProjectionCatchUpTrigger.php:
```diff
@@ -29,15 +38,80 @@ public function triggerCatchUp(Projections $projections): void
+
+        $attempts = 0;
+        while (!empty($queuedProjections)) {
+            usleep(random_int(100, 25000) + ($attempts * $attempts * 10)); // 50000μs = 50ms
```
I don't understand this line. Why is the time depending on the attempt – is that some exponential back off? And why based on a random value?

Maybe it would help to extract the loop to some static `exponentialBackOff()` function?

kitsunet replied:
Yes to both :)

I tried to make this more of an exponential backoff with an element of randomness to avoid any potential lockstep issues, will extract.

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Factory/ProjectionCatchUpTrigger/SubprocessProjectionCatchUpTrigger.php:
```diff
@@ -29,15 +38,80 @@ public function triggerCatchUp(Projections $projections): void
+            }
+
+            foreach ($queuedProjections as $key => $projection) {
+                if ($this->catchUpStatesCache->has($this->cacheKeyPrefix($projection) . 'QUEUED') === false) {
```
a rather minor detail, but I would suggest to remove the amount of magic keys by adding some
```php
private static function cacheEntryIdQueued(ProjectionInterface $projection): string {
  // ...
}
public static function cacheEntryIdRunning(ProjectionInterface $projection): string {
  // ...
}
```
the latter can be called from `SubprocessProjectionCatchUpCommandController::catchupCommand()`


Besides, it would make the logic more readable if we added dedicated methods for the cache interactions.
e.g. instead of
```php
if ($this->catchUpStatesCache->has($this->cacheKeyPrefix($projection) . 'RUNNING')) {
    return false;
}
$this->catchUpStatesCache->set($this->cacheKeyPrefix($projection) . 'RUNNING', 1);
```
something like
```php
if ($this->isCatchUpRunning($projection)) {
    return false;
}
$this->markCatchUpRunning($projection);
```

kitsunet replied:
Yeah, wasn't sure if static method in this place is _nice_ but I agree, I would be happier about that too.

Maybe I just make all of this a little class of it's own. I think that would be even nicer, `ProjectionCatchUpRunnerState($projection, $cache)` which then has stuff like ::isRunning() ::markRunning() ::markStopped()

kitsunet replied:
alright, update is on, looks much nicer now

bwaidelich reviewed (commented):
This is great already, thanks a lot for your efforts!
I find it a bit hard to understand the queueing logic (even though we came up with it together *g)
Maybe that can be formulated a bit more explicitly such that it is easier to find potential issues.
I added some first suggestions inline

kitsunet committed:
Update Neos.ContentRepositoryRegistry/Classes/Factory/ProjectionCatchUpTrigger/SubprocessProjectionCatchUpTrigger.php

Co-authored-by: Bastian Waidelich <b.waidelich@wwwision.de>

kitsunet committed:
TASK: Code cleanup and refactoring

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Command/SubprocessProjectionCatchUpCommandController.php:
```diff
@@ -31,7 +40,10 @@ class SubprocessProjectionCatchUpCommandController extends CommandController
     {
-        $contentRepository = $this->contentRepositoryRegistry->get(ContentRepositoryId::fromString($contentRepositoryIdentifier));
+        $contentRepositoryId = ContentRepositoryId::fromString($contentRepositoryIdentifier);
+        $runnerState = AsynchronousCatchUpRunnerState::create($contentRepositoryId, $projectionClassName, $this->catchUpStatesCache);
```
Wouldn't it make sense to make `AsynchronousCatchUpRunnerState` a singleton that already has the cache backend injected.
Otherwise we leak internal details and the exact cache name has to be specified everywhere this is used (and using a different cache will work, but not as expected).
Of course, as a consequence `$contentRepositoryId` and `$projectionClassName` would have to be arguments of the public methods then..
Alternatively we could create a factory and inject _that_:
```php
$this->catchUpRunnerStateFactory->build($contentRepositoryId, $projectionClassName)->setStopped();
```

nit pick: the `$runnerState` instance does not have to be created (and occupy memory) before the catchup is started

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Service/AsynchronousCatchUpRunnerState.php:
```diff
@@ -0,0 +1,60 @@
+/**
+ *
+ */
+final readonly class AsynchronousCatchUpRunnerState
```
I would suggest to skip the `Asynchronous` prefix since this central coordinator is also relevant for _synchronous_ catch-ups.
For example: I would consider switching my installations to sync mode (like @nezaniel did in his spaceship *g) for an additional performance boost. But with concurrent requests this could lead to the infamous "Failed to acquire checkpoint lock for subscriber '...' because it is acquired already" exception.


That again makes me think: Can we outsource some of the `SubprocessProjectionCatchUpTrigger` code so it can be reused for the sync mode (or others potentially) too?

bwaidelich reviewed (commented):
This looks so much cleaner and easier to comprehend, thank you!
Two more comments that I wasn't aware before..
Let me know if you could use a rubber duck :)

bwaidelich reviewed (commented):
@kitsunet I thought about this one again realized that `CATCHUPTRIGGER_ENABLE_SYNCHRONOUS_OPTION` is a bad naming choice since this is not so much about sync vs async but more about whether we expect multiple parallel threads or not.
For example: I would like to be able to simply switch the catch up trigger to the synchronous one for individual CRs for performance reasons – but they would still have to be queued in order to prevent concurrent processes from running into the "Failed to acquire checkpoint" exception.

I have a, somewhat radical, suggestion:
What if we replaced `AsynchronousCatchUpRunnerState` by some central authority that can be used to queue catchups independently from their implementation?

```php
final readonly class CatchUpQueue {

    public function __construct(
        private ContentRepositoryId $contentRepositoryId,
        private FrontendInterface $catchUpLock,
    ) {}

    public function queueCatchUp(ProjectionInterface $projection, \Closure $catchUpTrigger): void {
        // here goes your logic from SubprocessProjectionCatchUpTrigger and  AsynchronousCatchUpRunnerState
    }

    public function releaseCatchUpLock(string $projectionClassName): void {
        // to be called from the part that invokes ContentRepository::catchUp()
    }
}
```

I think that the change is smaller than it might first appear (since you already solved all of the logical issues) but it would allow us to use this for all implementations (for testing context we could consider replacing the cache with a `NullBackend`).

E.g. in `SubprocessProjectionCatchUpTrigger::triggerCatchUp()`:

```php
$catchUpQueue = $this->catchUpQueueFactory->build($contentRepositoryId);
foreach ($projections as $projection) {
    $catchUpQueue->queueCatchUp($projection, $this->startCatchUp(...));
}
```

Sorry for the forth and back, but I think we're getting there and I'm happy to help or take over!

bwaidelich commented:
The example code above is not correct yet because it would block catchups in the `foreach`, so maybe it has to be adjusted to:
```php
final readonly class CatchUpQueue {
     // ...
    public function queueCatchUp(array $projections, \Closure $catchUpTrigger): void {
        // keep looping over $projections until there are no more running remaining
    }
}
```

kitsunet commented:
Oh, that's a very cool idea, and I think this all makes sense, I'll see how far I get tomorrow and then we can see if you want to take over but at least I get your idea and know this part of the code already so it makes sense for me to go on.

kitsunet commented:
This is a though nut, I thought it would be nicer to wrap the debouncer - as I named it now - around the `ProjectionCatchUpTriggerInterface` also implementing the interface, this is nice as it just needs slight adjustments in `\Neos\ContentRepositoryRegistry\ContentRepositoryRegistry::buildProjectionCatchUpTrigger`, this seemed cleanest, but regardless of how we do it, we somehow need to "unlock" after a catchup is finished even in async. The clean way IMHO would be to use a CatchUpHook that triggers on `onAfterCatchUp` BUT that gets no arguments so I neither know which repository nor which projection was caught up :sadpanda: (I figured the repository is known to the factory creating the hook, so that is fine)
 I guess I will extend the interface but hat will make this a breaking change if someone implemented the interface...

kitsunet commented:
No, this also seems like a bad idea, because the hooks are per projection and I can't know from the outside which projection has said hook active. It feels unwise to have a _global_ debouncer but a per projection (and repository) unlock of it (via the hook).

SignalSlot _might_ be an option but then it becomes messy to get hold of the debouncer to unlock the state.

kitsunet commented:
I guess I would also like a rubber duck here :)

bwaidelich commented:
Using the `CatchUpHook` seems like a nice – although I did not fully understand why this is needed if we have a custom CLI handler to trigger the catch up.

> I guess I will extend the interface but hat will make this a breaking change if someone implemented the interface...

I think that's fine. It would only break for users that actually created a custom hook implementation – and it's really easy to fix

> SignalSlot might be an option

I would like the implementation to be independent from Flow as much as possible because something like this will be needed in other places, too..


> I guess I would also like a rubber duck here :)

I'm up for it – tomorrow?

kitsunet committed:
Rewrite to central deduplication queue

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/Projections.php:
```diff
@@ -93,4 +93,8 @@ public function getIterator(): \Traversable
     }
+    public function isEmpty(): bool
+    {
+        return count($this->projections) === 0;
```
Mini nitpick, this is _slightly_ faster and what we use in other places
```suggestion
        return $this->projections === [];
```

kitsunet replied:
I never really liked the readability of this (to me) but will change it.

bwaidelich replied:
for me it's the other way around, but it's probably a matter of habit, so nevermind!

bwaidelich commented on Neos.ContentRepository.Core/Classes/Projection/Projections.php:
```diff
@@ -93,4 +93,8 @@ public function getIterator(): \Traversable
     {
         yield from $this->projections;
     }
+    public function isEmpty(): bool
```
missing empty line :)
```suggestion

    public function isEmpty(): bool
```

kitsunet replied:
<img width="371" alt="Screenshot 2023-11-29 at 11 43 30" src="https://github.com/neos/neos-development-collection/assets/324408/a7c22ac9-7765-4cd0-bdfe-79c54a202c0d">

git integration let me astray :D

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Command/SubprocessProjectionCatchUpCommandController.php:
```diff
@@ -22,7 +23,14 @@ class SubprocessProjectionCatchUpCommandController extends CommandController
+     * @var VariableFrontend
+     */
+    protected VariableFrontend $catchUpStatesCache;
+
```
I wanted to suggest not to expose this non-API cache, but I realized that it's not even used, is it?
```suggestion
```

bwaidelich replied:
Ah, you already fixed that one :)

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Command/SubprocessProjectionCatchUpCommandController.php:
```diff
@@ -22,7 +23,14 @@ class SubprocessProjectionCatchUpCommandController extends CommandController
      * @var ContentRepositoryRegistry
      */
-    protected $contentRepositoryRegistry;
+    protected ContentRepositoryRegistry $contentRepositoryRegistry;
```
With that type hint, this is no longer a lazy injection.
Totally fine with me, but in that case I would suggest to replace it with a constructor injection for good measure :)

```php
public function __construct(
    private readonly ContentRepositoryRegistry $contentRepositoryRegistry,
) {
    parent::__construct();
}
```

kitsunet replied:
given that the controller has no other functionality it seems fine to me to not have it lazy. BUT I also was not aware that happens (it seems obvious though that it cannot be lazy), seems a rather big little pitfall :D

bwaidelich replied:
> seems a rather big little pitfall

Yes, I agree and that's why I try to avoid this everywhere.. It also makes the code pure PHP which feels cleaner (even though this class will always be a Flow class)

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Service/CatchUpDeduplicationQueue.php:
```diff
@@ -0,0 +1,171 @@
+    {
+        $queuedProjections = $this->triggerCatchUpAndReturnQueued($projections);
+        $attempts = 0;
+        /** @phpstan-ignore-next-line */
```
Just out of curiosity: What was PHPStan complaining about?

kitsunet replied:
the new `Projections::empty`is in an internal class  if you have suggestions I am happy, we could make Projections not internal or I could not use it, but it seems very convenient.

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Service/CatchUpDeduplicationQueue.php:
```diff
@@ -0,0 +1,171 @@
+use Neos\ContentRepository\Core\Projection\Projections;
+
+/**
+ *
```
Missing doc comments :)

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Service/CatchUpDeduplicationQueue.php:
```diff
@@ -0,0 +1,171 @@
+     */
+    public function releaseCatchUpLock(string $projectionClassName): void
+    {
+        $this->setStopped($projectionClassName);
```
`setStopped()` is only used here and contains a single line, I would suggest to inline that!?

kitsunet replied:
should be fine indeed!

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Service/CatchUpDeduplicationQueue.php:
```diff
@@ -0,0 +1,171 @@
+
+    private function triggerCatchUpAndReturnQueued(Projections $projections): Projections
+    {
+        $passToCatchUp = [];
```
I stumbled upon this name, what about `$projectionsToCatchUp`?

kitsunet replied:
right, that makes sense

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Service/CatchUpDeduplicationQueue.php:
```diff
@@ -0,0 +1,171 @@
+        while ($queuedProjections->isEmpty() === false) {
+            usleep(random_int(100, 100 + $attempts) + ($attempts * $attempts * 10));
+            $queuedProjections = $this->retryQueued($queuedProjections);
+            $attempts++;
```
There is no timeout / max attempts.. that makes me suspicious..

kitsunet replied:
I had that before but reasoning with myself says there shouldn't be any, we can't know how long the catchup takes, but we do know that we wanted another one, so we just got to wait...

kitsunet replied:
I did a quick test using the amp package to make this wait async (as in, other async stuff can happen while we wait), that works well, but I would do that in a follow up.

bwaidelich replied:
I used [react/promise](https://reactphp.org/promise/) for that in my custom implementation and I really like the simple way to use it:

```php
$this->catchUpTrigger->start()
    ->then(
        fn (SequenceNumber $sequenceNumber) => $this->foo($sequenceNumber),
        fn (Throwable $exception) => $this->bar($exception)
    );
```

(but AMP is probably just the same)

Still, I think it would be dangerous to kill the server if some projection hangs and would rather stop with a meaningful exception after a reasonable amount of time

kitsunet replied:
The question is, what is a reasonable amount of time, if a long catchup is running (eg after a big publish) it can take quite a while to get your queued catchup in. 

ASIDE that makes the bottom maybe useless, just came to my mind... What about this:
Now that we have this Queue also in the EventPersister to unlock after, we might as well defer the retriggering to after a catchup happened?! That way we could totally skip the loop here and just do something like

```
        $catchUp->run($eventStream);
        $catchUpHook?->onAfterCatchUp();
        $this->catchUpDeduplicationQueue->releaseCatchUpLock($projectionClassName);
        $this->catchUpDeduplicationQueue->startIfQueued($projectionClassName);
```
Means we only really block if we are forced by the `block()`  thing.

Note that you have two ways to get out of the loop, not only by starting another catchup there but also if it was dequeued due to another process starting a catchup, so IMHO it should be fine to risk, but we can obivously add a generous end condition, any suggestions? 1 minute ? 5?

kitsunet replied:
Means we only really block if we are forced by the `block()`  thing.

bwaidelich replied:
> we might as well defer the retriggering to after a catchup happened?!

We thought about this before, but this could mean that one process has to wait endlessly (not very likely, but in a scenario where many other systems interact with the CR).
That's why we decided to put the "burden" on to the invoking side of the new catch up (that now has to wait for the first catch up to finish)

kitsunet replied:
Right, good point! It's so tricky. I was almost writing that we only ever need to wait for a single queued catchup, but that is a fallacy of course as there can be a new one queued up while we run the previously queued one and nothing would take care of it if we only run a single catchup here.

Ok, back to the end condition. I think it should be pretty _high_ Reason beaing that otherwise you will get the timeout while waiting for .... error.

bwaidelich replied:
I agree, sth like 20 seconds I'd say.
Remember: This won't intercept the catchup or start a new one if the time exeeded. It just avoids that a process waits for other catchups endlessly

kitsunet replied:
Here we go, with a bit of explainer added as well.

bwaidelich replied:
I realized, that we can't use the PSR cache interfaces for this because they don't allow for exclusive cache entries. Instead we'll need some lock (e.g. [symfony/lock](https://symfony.com/doc/current/components/lock.html)) to prevent race conditions.
Thanks to the nice abstraction, it should be relatively easy to replace the cache.

I'll experiment a bit with the lock and ways to test this for reliability & concurrency and post an update here asap

kitsunet replied:
Yeah, this is not atomic, so it's not a 100% failsafe and for sure th more pressure the more duplicate runs probably can happen... I am not sure if it still isn't better than what we have now and we might merge this and refactor to lock or similar afterwards, but I am fine either way.

bwaidelich replied:
Note: I was on my smartphone and actually meant to comment on the PR itself, not on this particular code

> more pressure the more duplicate runs probably can happen

The problem is that the same event could be consumed multiple times by a single listener – it is currently prevented by our `CheckpointStorage` implementation, but I would really like to simplify that one.
I wrote a simplified example with a test using [paratest](https://github.com/paratestphp/paratest) and it failed every time – using the lock seemed faster and the race condition was gone.

There are also two more details that popped up.. Are you up to discuss those at some point?

bwaidelich reviewed (commented):
@kitsunet This looks awesome already, thanks for holding on!!

Beware: Most if not all of my inline comments are rather nitpicky – that's because I couldn't find any bigger issues ;)

Except for a two _potential_ ones:

## 1. Global Queue concept

With the CR registry we tried to avoid our "global service" thinking, but now all CRs would share the same `CatchUpDeduplicationQueue`, conceptually, i.e. sharing the same cache backend (my point: they could internally, but now they _have_ to)
That might not be an issue in practice, but if we were to add a factory like
```php
class CatchUpDeduplicationQueueFactory {

   // ...

   public function build(ContentRepositoryId $contentRepositoryId): CatchUpDeduplicationQueue
   {
        // ...
    }
}
```

And with that, maybe, _maybe_ it makes sense to take the one extra step and _do_ introduce an interface for it..

To be honest: While writing it down, I'm not sure what issue there could be.. Maybe we can just postpone this part

## 2. Tests

Annoying one, but since this is a crucial part for the system to work, I would suggest we add some tests.
We could at least add a functional test, creating two instances and see how they interact?
But the best would be a test that covers this with multiple threads. I made some great experience with [paratestphp/paratest](https://github.com/paratestphp/paratest) (we use that for the consistency tests for [neos/eventstore-doctrineadapter](https://github.com/neos/eventstore-doctrineadapter) for example).

I'm happy to help with this, and we can – of course – add those in a separate PR as well (to get this one in asap)

kitsunet committed:
Fix style issues

kitsunet committed:
Stop waiting for queued catchups after 20 seconds

bwaidelich commented:
A few more potential issues/considerations (in addition to the potential race condition mentioned above):

- Currently the queue is responsible for multiple subscribers which makes the code a bit hard to follow. One Queue per subscriber might simplify the implementation a lot (and provide more flexibility for future features like projection priorities etc)
- `ContentRepository::catchUpProjection()` currently ignores the lock (by definition because it is meant as the low-level trigger). That could be error prone if people would use this instead of the coordinated catchUp. Maybe we can turn things around to only allow for queueing the catch up via CR API and invoke the actual catchup via some internal API
- The lock is currently not released upon exceptions AFAIS

kitsunet commented:
>     * Currently the queue is responsible for multiple subscribers which makes the code a bit hard to follow. One Queue per subscriber might simplify the implementation a lot (and provide more flexibility for future features like projection priorities etc)

I should have tried this as I thought the same, but I _think_ that we might end up in a concurrent lock situation if we do not tackle the whole package at once (as then you end up in multiple while loops waiting for catchups to finish)

> 
>     * `ContentRepository::catchUpProjection()` currently ignores the lock (by definition because it is meant as the low-level trigger). That could be error prone if people would use this instead of the coordinated catchUp. Maybe we can turn things around to only allow for queueing the catch up via CR API and invoke the actual catchup via some internal API

We _could_ check if the lock is acquired in there and otherwise throw an exception (or just do nothing) ?


>     * The lock is currently not released upon exceptions AFAIS

Right that needs to be fixed for sure!

kitsunet committed:
Implement CatchUpDeduplication with locks

Using symfony/lock ensures atomic locking, which
should really prevent duplication even under load.

kitsunet commented:
Step one, replacing cache with lock, the overall logic is the same still.

kitsunet committed:
Split catchUp queue and run lock

kitsunet commented:
We now have two separate lock steps, a "run" lock around the catchUp of a specific projection, preventing that same projection catchUp to be run multiple times in parallel, and the deduplication "queue" lock that will check if a catchUp is already running and create a queue lock to prevent multiple processes to wait for another catchup run and all having to potentially spawn background processes. This second queue lock is not atomic as it depends on the "run" lock. This should be fine, it's there to prevent runaway spawning of parallel subrequests in busy installation and for that it should work fine with it's randomized back off.

kitsunet committed:
Fix style issues

mhsdesign commented on Neos.ContentRepositoryRegistry/Tests/Unit/Service/CatchUpDeduplicationQueueTest.php:
```diff
@@ -0,0 +1,76 @@
+/**
+ * This is not a regular unit test, it won't run with the rest of the testsuite.
+ * the following two commands after another would run the parallel tests and then the validation of results:
+ * requires "brianium/paratest": "^6.11"
```
i see because the latest para test latest requires phpunit 10 but flow blocks that.

mhsdesign commented:
I think this is now superseded by https://github.com/neos/neos-development-collection/pull/5321
